streamingest: add replication states for `init` and `cutover`

lidorcarmel opened the pull request:
Adding 2 new tenant states: `INITIALIZING REPLICATION ` and `REPLICATION CUTTING OVER`.

Adding the cutover time to the `SHOW TENANT WITH REPLICATION` command.

See the commits for more info.

Fixes: #94707

Epic: CRDB-18752

Release note: None

Commit:
streamingest: add INITIALIZING REPLICATION tenant state

Until the initial scan is completed we cannot cutover to the secondary
tenant - the state of that tenant will be, with this PR,
`INITIALIZING REPLICATION`. The tenant can stay in that state for some
time.

Note that with this change the replicated tenant cannot be in the `ADD`
state.

Epic: CRDB-18752

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/94011)
<!-- Reviewable:end -->

lidorcarmel commented on pkg/sql/show_tenant.go:
```diff
@@ -92,6 +94,45 @@ func (n *showTenantNode) getTenantName(params runParams) (roachpb.TenantName, er
 	return roachpb.TenantName(*name), nil
 }
 
+func (n *showTenantNode) initReplicationStats(
```
this block is just a code move.

lidorcarmel commented on pkg/sql/show_tenant.go:
```diff
@@ -101,75 +142,50 @@ func (n *showTenantNode) startExec(params runParams) error {
-					tenantName, jobId)
+
+		switch job.Status() {
+		case jobs.StatusPending, jobs.StatusRunning, jobs.StatusPauseRequested:
```
the next few lines are the main change in this pr.

Commit:
streamingest: add a cutover replication state

The destination tenant that we're replicating to can be, with this
commit, in `REPLICATION CUTTING OVER` state. This means that we have a
cutover timestamp and that the tenant is in the process of finishing
replication and switching to the `ACTIVE` state.

This commit also adds the cutover timestamp to the `SHOW TENANT WITH
REPLICATION` command. Note that the same timestamp is already printed
when running the ALTER TENANT command to COMPLETE replication (do the
cutover).

Epic: CRDB-18752

Release note: None

lidorcarmel commented:
and note that the first 3 commits are from PR #93755

adityamaru commented on pkg/ccl/streamingccl/streamingest/replication_stream_e2e_test.go:
```diff
@@ -903,6 +903,14 @@ func TestTenantStreamingShowTenant(t *testing.T) {
 	defer cleanup()
 	producerJobID, ingestionJobID := c.StartStreamReplication(ctx)
 
+	rowStr := c.DestSysSQL.QueryStr(t, fmt.Sprintf("SHOW TENANT %s WITH REPLICATION STATUS", args.DestTenantName))
```
Do you want to port this test over to the datadriven framework now? I also prefer if we had a deterministic test of the different states. I think we can set `stream_replication.job_checkpoint_frequency` to 0 to disable all progress checkpoints. Then we can check that its in initReplication. We can then set the frequency to non-zero and ensure its in replicating and so on?

lidorcarmel replied:
> Do you want to port this test over to the datadriven framework now?

that was the plan but later I thought I want to keep this test.

> I think we can set stream_replication.job_checkpoint_frequency to 0

I'll looks at that, I think this is what I was looking for, thanks!

lidorcarmel replied:
> I think we can set stream_replication.job_checkpoint_frequency to 0

unfortunately this would not help here, because we will still populate the lag info, even when setting that to 0.
let me think about this some more, and we can also discuss offline.

adityamaru commented on pkg/ccl/streamingccl/streamingest/replication_stream_e2e_test.go:
```diff
@@ -929,9 +929,10 @@ func TestTenantStreamingShowTenant(t *testing.T) {
 		jobId         int
 		maxReplTime   time.Time
 		protectedTime time.Time
+		cutoverTime   []byte // should be nil
```
Can we add a test that once a cutover is initiated, this isn't nil

lidorcarmel replied:
I'll add that to the datadriven.

lidorcarmel replied:
sorry, we already test that in the datadriven test, right?

adityamaru replied:
oh I meant that in addition to checking its nil when there's no cutover if we can ensure that cutover actually shows up in the output of `SHOW TENANT` once we have issued a `COMPLETE REPLICATION`. I'll leave it up to you if it seems necessary.

lidorcarmel replied:
good point, added at the end of this test.

adityamaru commented on pkg/sql/show_tenant.go:
```diff
@@ -171,7 +172,12 @@ func (n *showTenantNode) startExec(params runParams) error {
 			} else {
-				n.tenantStatus = replicating
+				progress := n.replicationInfo.IngestionProgress
+				if progress != nil && !progress.CutoverTime.IsEmpty() {
```
If this cutover time is far in the future, our behaviour is going to be to allow the replication to reach that timestamp before actually cutting over. For all that time we will be in state `REPLICATION CUTTING OVER`, does this seem strange to others? Should we instead only show this status when we're in `maybeRevertToCutoverTimestamp`?

lidorcarmel replied:
I don't think we should be in `REPLICATING` when a cutover timestamp is set, because this means that a user runs `ALTER TENANT .. COMPLETE ..` and the status of the tenant doesn't change.
if we want we can have another status such as `CUTOVER PENDING` but not sure it's really important now.

adityamaru replied:
I like `REPLICATION PENDING CUTOVER` or something to that effect, I think it may cause alarm to the user if they see `REPLICATION CUTTING OVER` when they explicitly set a time an hour in the future from the replicated time.

adityamaru commented on pkg/sql/show_tenant_test.go:
```diff
@@ -0,0 +1,103 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sql
```
do you want to move `TestTenantStreamingShowTenant` here as well?

lidorcarmel reviewed (commented):
thanks!!
